### ISSUE TITLE
Kill a process when spring stop

### ISSUE THREAD

**Issue by GCorbel** [rails/spring#424]
This question is related to [this issue](https://github.com/rwz/ember-cli-rails/issues/203) and [this question](http://stackoverflow.com/questions/32226212/how-to-stop-a-process-when-another-is-stopped).

I'm looking for a way to kill Ember, or any process, when spring is stopped. Is it possible. Do you have an idea?

**Comment by grosser** [comment]
allowing to overwrite pre-boot steps like start & stop might solve that ... maybe some `config/spring-preboot.rb` could be supported where you can go crazy and make stop do all kinds of things ...

**Comment by GCorbel** [comment]
Monkey patching? No cleaner way?

**Comment by GCorbel** [comment]
Use `Signal.trap("TERM") { exec('touch /tmp/test123') }` seems to works when spring is stopped. I don't know if their can be conflicts.

**Comment by GCorbel** [comment]
I think I can you only one `Signal.trap` in the whole app so it's not a good way to go. Monkey Patching does not work because Spring is loaded after the initializers (I think) and methods I'm trying to create are overwritten. Any clue?

**Comment by grosser** [comment]
FYI You can trap a signal and then call the previous trap:

https://github.com/grosser/parallel/blob/master/lib/parallel.rb#L158-L171

but not something I'd recommend ...

some pre-boot configuration file would be nice to be able to control
start/stop etc behavior ...

**Comment by GCorbel** [comment]
I agree but I don't know how to do. I have to do something when spring stops so I have to know when it stop and interact with EmberCLI. I don't see a solution to stop a process when spring is stopped. Do you have a clue?

**Comment by GCorbel** [comment]
@grosser, the only think I see is to do a callback function `Spring.after_stop` or `Spring.after_kill`. If you agree, I will do a pull request.

**Comment by grosser** [comment]
https://github.com/rails/spring/pull/426 would allow you to do:

```Ruby
Spring::Client::Stop.prepend(Module.new do
  def call(args)
    super
    # do something else
  end
end)
```

does that look good ?

**Comment by GCorbel** [comment]
It does not work. Spring::Client is undefined so I required it and it's never triggered.

**Comment by grosser** [comment]
did you put it in config/spring_preboot.rb and use gem installed from the branch ? (`rake install`)

**Comment by GCorbel** [comment]
I did not. Now, it installed it properly. The code is called for each run even if I did not `spring stop`.

**Comment by GCorbel** [comment]
Sorry, I was wrong. It seems to work. I have some verifications to do before to close this issue.

**Comment by grosser** [comment]
Once you verified I'll merge the PR and release a new version :)

**Comment by GCorbel** [comment]
I only have two problems but about EmberCLI-Rails. First, EmberCLI[:frontend] is undefined on `spring_preboot.rb`. I suppose all gems are not loaded.

The second problem, too linked to EmberCLI-Rails I think, is than there is 3 processes launched. The PID stored in the app is the first one and I need to kill the second one. I have to work to resolve this issue.

**Comment by grosser** [comment]
can you `ps -ef | grep foobar` and then get the pids to kill them ?
but the preboot.rb looks like it helps / should I merge it ?

**Comment by GCorbel** [comment]
With `ps -ef | grep ember`, I can have other processes not related.

`preboot.rb` will help for sure. You can merge it.
